mantidproject/mantid
# Improve ellipsoid shape estimation and add translation to ellipsoid shape

**Description of work.**

The covariance matrix is estimated iteratively (excluding points beyond 3 stdevs in each iteration). In addition there is now an option to perform the integration around the estimated centroid (this involved adding a translation property to the Ellipsoid peak shape and allowing for this to be plotted in the Sliceviewer) - this helps catch the tail of the peak to low Q.
Have also made a performance improvement by adding a rough test to see if MDBox is definitely not in peak region (and therefore avoid going through the events of such boxes).

Here is demo of 3 iterations of covariance determination on some WISH data (spherical region indicated in yellow).
![image](https://user-images.githubusercontent.com/55979119/109624767-7ec19b80-7b36-11eb-857c-837cf76d5223.png)


**To test:**

**NOTE: you will get an error when integrating the fake data because it doesn't have the instrument definition when it tries to check the peak is fully on the detector - it will still perform the integration.**

*Centroid feature*
(1) Run this script without error.

```
ws = CreateMDWorkspace(Dimensions='3', Extents='0,2,0,2,0,2',
                       Names='Q_lab_x,Q_lab_y,Q_lab_z', Units='U,U,U',
                       Frames='QLab,QLab,QLab',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
FakeMDEventData(ws, EllipsoidParams='1e+04,1.2,1,1,1,0,0,0,1,0,0,0,1,0.01,0.04,0.02,1', RandomSeed='3873875')

# add peak
p = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
SetUB('peaks', 1,1,1,90,90,90)
pk = p.createPeak([1,1,1]) # axes aligned with viewing axes
p.addPeak(pk)
# integrate with bg
IntegratePeaksMD(InputWorkspace='ws', PeakRadius=0.6, BackgroundInnerRadius=0.6,
    BackgroundOuterRadius=1, PeaksWorkspace='peaks', OutputWorkspace='peaks_int', 
    Ellipsoid=True, UseOnePercentBackgroundCorrection=False)
# integrate with bg use centroid
IntegratePeaksMD(InputWorkspace='ws', PeakRadius=0.6, BackgroundInnerRadius=0.6,
    BackgroundOuterRadius=1, PeaksWorkspace='peaks', OutputWorkspace='peaks_int_centroid', 
    Ellipsoid=True, UseCentroid=True, UseOnePercentBackgroundCorrection=False)
```

(2) This is an extreme example where the peak nominally at QLab = (1,1,1) is placed at (1.2, 1,1) - but you should see 
- The peak ws with the centroid as integration centre has shifted to the peak centre
-  In addition the shape ellispoid produced with UseCentroid=False should be larger and not fit the peak shape well (this is because using the nominal position increases the variance)

![image](https://user-images.githubusercontent.com/55979119/109656807-b6dbd500-7b5c-11eb-8752-fd6f604ee302.png)

(3) Stop scaling ellipsoid major axis to be equal to peak radius

```
# integrate with bg use centroid but don't scale major axis
IntegratePeaksMD(InputWorkspace='ws', PeakRadius=0.6, BackgroundInnerRadius=0.6,
    BackgroundOuterRadius=1, PeaksWorkspace='peaks', OutputWorkspace='peaks_int_centroid_unfixed', 
    Ellipsoid=True, UseCentroid=True, FixMajorAxisLength = False, UseOnePercentBackgroundCorrection=False)
```
- The 'unfixed' ellipsoid should be smaller but still contain the peak
![image](https://user-images.githubusercontent.com/55979119/109657893-e6d7a800-7b5d-11eb-985d-a93d84356eee.png)

*Iterative covariance determination*

(4) Add some uniform background and perform integration with 1 and 2 iterations
```
FakeMDEventData(ws, UniformParams='1e4', RandomSeed='3873875')
for n in range(1,3):
    IntegratePeaksMD(InputWorkspace='ws', PeakRadius=0.6, BackgroundInnerRadius=0.6,
        BackgroundOuterRadius=1, PeaksWorkspace='peaks', OutputWorkspace='peaks_int_centroid_unfixed_n{0}'.format(n), 
        Ellipsoid=True, UseCentroid=True, FixMajorAxisLength = False, MaxIterations = n,
        UseOnePercentBackgroundCorrection=False)
```
![image](https://user-images.githubusercontent.com/55979119/109658580-aaf11280-7b5e-11eb-91c4-44bb0d647ea4.png)




Fixes #30573 and #30575






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
